design/hostclaim-multitenancy.md

```diff
@@ -0,0 +1,582 @@
+  kind: HostClaim
+  metadata:
+    name: my-host
+  spec:
+    online: true
+    image:
+      checksum: https://url_image.qcow2.md5
+      url: https://url_image.qcow2
+      format: qcow2
+    userData:
+      name: my-user-data
+    networkData:
+      name: my-network-data
+    hostSelector:
+      matchLabels:
+        infra-kind: medium
+  ```
+
+* The workload is launched. When the machine is fully provisioned, the boolean
+  field ready in the status becomes true. I can stop the server by changing
+  the online status. I can also perform a reboot by targeting specific
+  annotations in the reserved ``host.metal3.io`` domain.
+* When I destroy the host, the association is broken and another user can take
+  over the server.
+
+##### Server Administrator Point of View
+
+As the owner of the bare-metal servers, I would like to easily control the users
+that can use my servers and control the information I share about the servers.
+
+In each namespace hosting BareMetalHosts, I define one
+or several HostDeployPolicy resources. Each resource defines a filter on the
+namespaces of the HostClaims in the field ``hostclaimNamespaces``.
+
+The specification of the HostDeployPolicy contains one or several of the
+optional fields:
+
+* ``namespaces``: the content of this field is a list of namespaces. The name of
+  the namespace where the HostClaim is must be a member of the list.
+* ``nameMatches``: the content of this field is a regular expression
+  (in Golang re2 syntax). The name of the namespace where the HostClaim is located
+  must match this regular expression.
+* ``withLabels``: the content of the field is a list of strings representing
+  labels. The namespace of the HostClaim must have a label belonging to the
+  list.
+
+To successfully match, the namespace must fulfill the constraints of all the
+fields that are defined. An empty HostDeployPolicy authorizes all namespaces.
+A HostClaim is authorized to bind a BareMetalHost if and only if the namespace
+of the HostClaim is successfully filtered by at least one HostDeployPolicy in
+the namespace of the BareMetalHost. In that case, the binding between the
+HostClaim and the BareMetalHost is authorized.
+
+A HostDeployPolicy specification also contains a field ``metadataCopy``. It defines
+rules for which annotations are copied (``annotationsMatch``) and which
+labels are copied (``labelsMatch``). Both fields contain regular expressions.
+
+For example,
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: HostDeployPolicy
+metadata:
+  name: policy
+  namespace: infra
+spec:
+  hostclaimNamespaces:
+    namespaces:
+    - cluster1
+    - cluster2
+  metadataCopy:
+    annotationsMatch: 'mydomain/.*'
+    labelsMatch: 'selectors/.*'
+```
+
+will make the BareMetalHosts in namespace ``infra`` to HostClaims in namespaces
+``cluster1`` and ``cluster2``. All labels in domain ``selectors`` will be
+available for selection as annotations in ``mydomain`` for constructing
+cloud-init ``metadata``.
+
```

Comment:
Missing part: how is binding between BMH and HostClaim established? Via `consumerRef`?

Reply:
On the first comment, I remove this part. The identityRef is on the Metal3Machine NOT on the hostClaim. The yaml does not bring anything but confusion. I have added a sentence explictly telling this. We may revisit it later to address the case of a workload not managed by cluster-api-provider-metal3 that we would like to execute on a remote cluster (but is it really worth the complexity as we still need to read HardwareData on the remote cluster).

Reply:
If you don't have any back link from the BMH to the HC, your controller won't be able to receive reconciliation requests for the HC whenever the BMH changes, which is crucial e.g. for tracking the deployment process. I think you need some link, which is probably `consumerRef`.

Reply:
Yes it is the consumerRef. line 152-153 should tell that may be not at the right place.